core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -1900,6 +1902,60 @@ private static class GreatestConvertlet implements SqlRexConvertlet {
     }
   }
 
+  /** Convertlet that converts {@code GREATEST} and {@code LEAST}. */
+  private static class GreatestPgConvertlet implements SqlRexConvertlet {
+    @Override public RexNode convertCall(SqlRexContext cx, SqlCall call) {
```

Comment:
this expansion is quadratic in the list size.
is this something we should worry about?

Reply:
I based this on the existing `GreatestConvertlet` class. It converts in a very similar way. The original `GreatestConvertlet` was written 8 years ago, so a different solution may make more sense now.

In general, the user calls the function with a list of scalars, such as `GREATEST(1, 2, 3, 4)`. There must be an upper limit, since it must fit in a SQL call that can be parsed.

Reply:
There are some PRs attempting to solve https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-3094, which happens when a table has too many columns. So what we assume as a small constant can actually be big.

I could see cases where compiler-generated code could produce a long list as an argument for GREATEST. 
So I wonder whether throwing an exception for a very long list of arguments (e.g., >100) would make sense.

Reply:
It would be preferable to give a useful error message to the caller, such as "too many arguments". I don't have a good idea what would make sense for the limit.

I tested `CONCAT` in PostgreSQL and it gives an error after 100 arguments. It probably makes sense to bring that limitation into Calcite.

I also tested `GREATEST` in PostgreSQL. PostgreSQL must do something clever with the arguments, since it works up to at least 1001 arguments.

Reply:
The problem is that you want to compile to sql. In you use a language with local variables you can do it in size O(N). And postgres is not constrained to compile to sql.

Reply:
An expression like `GREATEST(1, 2, ...)` is compiled and becomes a `SqlCall` object with arguments. The convertlet will then replace the `SqlCall` object with a new `SqlCall` for `CASE`. We are only manipulating objects after they are parsed, we do not generate SQL text and parse it again.

This will still explode original expression and use potentially much more memory.

With simple integer arguments, this PR works with at least 1000 values.

If no limit is in place, the user is still restricted by the max query length and what will fit in memory after the convertlet is run.

Reply:
yes, your intermediate language is as expressive as SQL. That's what I mean.
But if you have let expressions you can do much better.

Reply:
You are right. Let expressions would avoid repeated evaluation of the arguments. Does Calcite support LET expressions?